src/main/java/org/opensearch/security/privileges/PitAccessEvaluator.java

```diff
@@ -0,0 +1,344 @@
+        if(pitInfos.isEmpty()) {
+            presponse.allowed = true;
+            presponse.markComplete();
+        }
+        List<String> pitIds = new ArrayList<>();
+        pitIds.addAll(pitInfos.stream().map(ListPitInfo::getPitId).collect(Collectors.toList()));
+        Map<String, String[]> pitToIndicesMap = OpenSearchSecurityPlugin.GuiceHolder.getPitService().getIndicesForPits(pitIds);
+        Map<String, ListPitInfo> pitToPitInfoMap =  new HashMap<>();
+
+        for(ListPitInfo pitInfo : pitInfos) {
+            pitToPitInfoMap.put(pitInfo.getPitId(), pitInfo);
+        }
+        List<ListPitInfo> permittedPits = new ArrayList<>();
+        for (String pitId : pitIds) {
+            String[] indices = pitToIndicesMap.get(pitId);
+            HashSet<String> indicesSet = new HashSet<>(Arrays.asList(indices));
+
+            final ImmutableSet<String> INDICES_SET = ImmutableSet.copyOf(indicesSet);
+            final IndexResolverReplacer.Resolved pitResolved =
+                    new IndexResolverReplacer.Resolved(INDICES_SET, INDICES_SET, INDICES_SET,
+                            ImmutableSet.of(), SearchRequest.DEFAULT_INDICES_OPTIONS);
+
+            final Set<String> allPermittedIndices = securityRoles.reduce(pitResolved,
+                    user, new String[]{action}, resolver, clusterService);
+            if(isDebugEnabled) {
+                log.debug("Evaluating PIT ID : " + pitId );
+            }
+            if (allPermittedIndices.size() == INDICES_SET.size()) {
+                if(isDebugEnabled) {
+                    log.debug(" Permitting PIT ID : " + pitId);
+                }
+                permittedPits.add(pitToPitInfoMap.get(pitId));
+            }
+        }
+        if (permittedPits.size() > 0) {
+            ((GetAllPitNodesRequest) request).setGetAllPitNodesResponse(new GetAllPitNodesResponse(permittedPits,
+                    ((GetAllPitNodesRequest) request).getGetAllPitNodesResponse()));
+            presponse.allowed = true;
+            presponse.markComplete();
+        }
+        return presponse;
+    }
+
+    /**
+     * Handle access for 'delete all PITs' operation
+     */
+    private PrivilegesEvaluatorResponse handleDeleteAllPitAccess(DeletePitRequest deletePitRequest, ClusterService clusterService,
+                                          User user, SecurityRoles securityRoles, final String action,
+                                          IndexNameExpressionResolver resolver,
+                                          PrivilegesEvaluatorResponse presponse) throws InterruptedException {
+        List<String> permittedPits = new ArrayList<>();
+        List<String> pitIds = getAllPitIds();
+        // allow delete pit operation if there are no pits in the cluster ( response should be empty )
+        if(pitIds.isEmpty()) {
+            deletePitRequest.clearAndSetPitIds(pitIds);
+            presponse.allowed = true;
```

Comment:
The authorization on a request should not be decided by whether there are PITs in a cluster. It should always be evaluated with permissions instead. If the user does not have right permissions, the request should be denied no matter whether there are PITs in a cluster or how many PITs there are.

Reply:
Same comment as above, we do evaluation of permissions, if user does not have permission, we'll not reach to this custom code.